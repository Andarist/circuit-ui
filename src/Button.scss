@import "./shared";

.btn {
  transition: color .2s cubic-bezier(.4, .0, .2, 1), background-color .2s cubic-bezier(.4, .0, .2, 1);
  display: inline-block;
  border: 1px solid;
  border-radius: $border-radius;
  cursor: pointer;
  font-size: $font-size-l;
  font-weight: 500;
  padding: 11px $spacing-xl;
  position: relative;
  white-space: nowrap;
  text-align: center;
  text-decoration: none;
  overflow: hidden;
  vertical-align: middle;
  text-overflow: ellipsis;
  line-height: $line-height-s;

  &:hover,
  &:active,
  &:focus {
    text-decoration: none;
    outline: none;
  }

  &:hover,
  &:active {
    color: $color-brand-darker;
  }

  &[disabled], &.btn--disabled {
    opacity: .4;
    cursor: default;
    pointer-events: none;
  }

  &--loading {
    transition: padding .2s ease-in-out;
    padding-right: $spacing-xxl !important;

    .spinner {
      display: inline-block;
      margin-left: $spacing-xs;
    }
  }

  // TODO: remove the highlight modifier as we're
  //       going to stick with cta as terminology.
  &#{&}--cta, &#{&}--highlight {
    background-color: $color-highlight;
    border-color: $color-highlight;
    color: white;

    &:hover,
    &:active,
    &:focus {
      background-color: $color-highlight-darker;
      border-color: $color-highlight-darker;

      &[disabled], &.btn--disabled {
        background-color: $color-highlight;
        border-color: $color-highlight;
      }
    }
  }

  // TODO: remove the outline modifier as we're
  //       going to stick with secondary as terminology.
  &#{&}--secondary, &#{&}--outline {
    background-color: transparent;
    border-color: $color-highlight;
    color: $color-highlight;

    &:hover,
    &:active,
    &:focus {
      border-color: $color-highlight-darker;
      color: $color-highlight-darker;

      &[disabled], &.btn--disabled {
        border-color: $color-highlight;
        color: $color-highlight;
      }
    }
  }

  // TODO: remove the outline-dark modifier as we're going
  //       to strick with secondary as terminology.
  &#{&}--secondary-dark, &#{&}--outline-dark {
    background-color: transparent;
    border: 1px solid $color-dark;
    color: $color-dark;

    &:hover {
      color: white;
      background-color: $color-dark;
    }
  }

  &#{&}--secondary-white, &#{&}--outline-white {
    background-color: transparent;
    border: 1px solid #FFF;
    color: #FFF;

    &:hover {
      color: $color-error;
      background-color: #FFF;
    }
  }

  // TODO: clarify if this will stay. Currently it's
  //       only used by the transaction filters on
  //       mobile.
  &#{&}--grey {
    background-color: $color-light;
    border-color: $color-light;
    color: $color-dark;

    &:hover,
    &:active,
    &:focus {
      background-color: darken($color-light, 5%);
      border-color: darken($color-light, 5%);

      &[disabled], &.btn--disabled {
        background-color: $color-light;
        border-color: $color-light;
      }
    }
  }

  &#{&}--warning {
    background-color: darken($color-warning, 5%);
    border-color: darken($color-warning, 5%);
    color: white;

    &:hover,
    &:active,
    &:focus {
      background-color: darken($color-warning, 5%);
      border-color: darken($color-warning, 5%);
    }
  }

  &#{&}--alert {
    background-color: $color-error;
    border-color: $color-error;
    color: white;

    &:hover,
    &:active,
    &:focus {
      background-color: $color-error;
      border-color: $color-error;
    }
  }

  &#{&}--s {
    padding: $spacing-xs $spacing-m;
    font-size: $font-size-m;
    line-height: $line-height-s;

    &.btn--loading .spinner {
      right: 9px;
      top: 10px;
    }
  }
}

/* Text buttons */

.text-btn {
  background: none;
  border: 0;
  cursor: pointer;
  font-size: $font-size-l;
  font-weight: 500;
  line-height: $line-height-s;
  padding: 0;

  &[disabled],
  &#{&}--disabled {
    cursor: default;
    pointer-events: none;
  }

  &#{&}--highlight {
    color: $color-highlight;

    &[disabled], &.btn--disabled {
      color: lighten($color-highlight, .2);
    }
  }

  &#{&}--brand {
    color: $color-brand;

    &[disabled], &.btn--disabled {
      color: lighten($color-brand, 20%);
    }
  }

  &#{&}--dark {
    color: $color-dark;

    &[disabled], &.btn--disabled {
      color: lighten($color-dark, 20%);
    }
  }

  &:focus { outline: none }
}

