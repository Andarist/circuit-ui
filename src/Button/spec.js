import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Button from '.';

describe('Button', () => {
  it('should not render if there happens to be no click handler, label, or children', () => {
    const button = renderer.create(<Button />);
    expect(button).toMatchSnapshot();
  });
  it('should be disabled if the disabled property is passed', () => {
    const button = renderer.create(
      <Button disabled={true} onClick={() => {}}>
        I am disabled
      </Button>
    );
    expect(button).toMatchSnapshot();
  });
  it('should take the body text as a child', () => {
    const output = mount(<Button onClick={() => {}}>Hello World</Button>);
    expect(output.text()).toContain('Hello World');
  });
  it('should respect label or body', () => {
    const button = renderer.create(
      <Button label="My label" onClick={() => {}} />
    );
    expect(button).toMatchSnapshot();
  });
  it('should accept classnames as props', () => {
    const button = renderer.create(
      <Button className="btn--loading" onClick={() => {}}>
        I might be loading
      </Button>
    );
    expect(button).toMatchSnapshot();
  });
});
