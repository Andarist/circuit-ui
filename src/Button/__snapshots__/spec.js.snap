// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should accept classnames as props 1`] = `
<button
  className="btn btn--loading"
  onClick={[Function]}
  title="I might be loading"
>
  I might be loading
</button>
`;

exports[`Button should be disabled if the disabled property is passed 1`] = `
<button
  className="btn btn--disabled"
  onClick={[Function]}
  title="I am disabled"
>
  I am disabled
</button>
`;

exports[`Button should not render if there happens to be no click handler, label, or children 1`] = `null`;

exports[`Button should respect label or body 1`] = `
<button
  className="btn"
  onClick={[Function]}
  title="My label"
>
  My label
</button>
`;
